java-checks/src/main/java/org/sonar/java/checks/regex/DuplicatesInCharacterClassCheck.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2020 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks.regex;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.TreeMap;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import org.sonar.check.Rule;
+import org.sonar.java.regex.RegexParseResult;
+import org.sonar.java.regex.ast.CharacterClassUnionTree;
+import org.sonar.java.regex.ast.CharacterRangeTree;
+import org.sonar.java.regex.ast.PlainCharacterTree;
+import org.sonar.java.regex.ast.RegexBaseVisitor;
+import org.sonar.java.regex.ast.RegexTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+
+@Rule(key = "S5869")
+public class DuplicatesInCharacterClassCheck extends AbstractRegexCheck {
+
+  private static final String MESSAGE = "Remove duplicates in this character class.";
+
+  @Override
+  public void checkRegex(RegexParseResult regexForLiterals, MethodInvocationTree mit) {
+    if (!regexForLiterals.hasSyntaxErrors()) {
+      DuplicateFinder duplicateFinder = new DuplicateFinder();
+      duplicateFinder.setActiveFlags(getFlags(mit));
```

Comment:
I would prefer to have this logic within the constructor of the `DuplicateFinder`, so we could rewrite this whole block such as:
```
if (!regexForLiterals.hasSyntaxErrors()) {
  regexForLiterals.getResult().accept(new DuplicateFinder(mit));
}

...

private class DuplicateFinder extends RegexBaseVisitor {
  private DuplicateFinder(MethodInvocationTree mit) {
    setActiveFlags(getFlags(mit));
  }

  ...
}

```

Or, what about having a final method `setActiveFlags(MethodInvocationTree mit)` within the `RegexBaseVisitor`, which would itself calls the currently private `getFlags(mit)` ?
So in our case we could just call `myVisitor.setActiveFlags(mit)`. I have the feeling that we will always have to call getFlags and setActiveFlags together.

Reply:
As discussed, I'll take care of this later in another PR after related changes from #3033 are merged.